### PR TITLE
Server whitelist

### PR DESCRIPTION
**Pull request by sclassen**

### CODE REVIEW

File: core/src/main/java/net/adoptopenjdk/icedteaweb/client/controlpanel/panels/ServerWhitelistPanel.java [hunk 1]
```diff
@@ -46,9 +56,97 @@ public ServerWhitelistPanel(final DeploymentConfiguration config) {
...
+                        setIcon(icon);
+                    } else {
+                        setIcon(null);
+                    }
+                } else {
+                    setText(null);
+                    setIcon(null);
+                }
+                if (isSelected) {
+                    setBackground(table.getSelectionBackground());
```
Comment (hendrikebbers) [inline core/src/main/java/net/adoptopenjdk/icedteaweb/client/controlpanel/panels/ServerWhitelistPanel.java:88]: is this background / foreground setting really needed? I assume the default renderer already does this

File: core/src/main/java/net/adoptopenjdk/icedteaweb/client/controlpanel/panels/ServerWhitelistPanel.java [hunk 2]
```diff
@@ -46,9 +56,97 @@ public ServerWhitelistPanel(final DeploymentConfiguration config) {
...
+        table.setFillsViewportHeight(true);
+        table.setAutoResizeMode(JTable.AUTO_RESIZE_ALL_COLUMNS);
+        table.setAutoCreateRowSorter(true);
+
+        final TableColumnModel colModel = table.getColumnModel();
+        colModel.getColumn(0).setPreferredWidth(100);
+        colModel.getColumn(1).setPreferredWidth(250);
+
+        final ImageIcon icon = SunMiscLauncher.getSecureImageIcon("net/sourceforge/jnlp/resources/warn16.png");
+        final DefaultTableCellRenderer cellRenderer = new DefaultTableCellRenderer() {
```
Comment (hendrikebbers) [inline core/src/main/java/net/adoptopenjdk/icedteaweb/client/controlpanel/panels/ServerWhitelistPanel.java:72]: maybe we externalise this class

File: core/src/main/java/net/adoptopenjdk/icedteaweb/client/controlpanel/panels/ServerWhitelistPanel.java [hunk 3]
```diff
@@ -46,9 +56,97 @@ public ServerWhitelistPanel(final DeploymentConfiguration config) {
...
+            @Override
+            public Object getValueAt(int rowIndex, int columnIndex) {
+                WhitelistEntry whitelistEntry = whitelist.get(rowIndex);
+                switch (columnIndex) {
+                    case 0:
+                        return whitelistEntry.getRawWhitelistEntry();
+                    case 1:
+                        return new WhitelistEntryState(whitelistEntry);
+                    default:
+                        throw new IllegalArgumentException();
```
Comment (hendrikebbers) [inline core/src/main/java/net/adoptopenjdk/icedteaweb/client/controlpanel/panels/ServerWhitelistPanel.java:131]: Please add some text

File: core/src/main/java/net/sourceforge/jnlp/util/whitelist/WhitelistEntry.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+            return false;
+        } else {
+            return isUrlProtocolMatching(effectiveWhitelistEntry, url) && isUrlHostMatching(effectiveWhitelistEntry, url) && isUrlPortMatching(effectiveWhitelistEntry, url);
+        }
+    }
+    private static boolean isUrlProtocolMatching(URL url1, URL url2) {
+        return Objects.equals(url1.getProtocol(), url2.getProtocol());
+    }
+
+    private static boolean isUrlHostMatching(URL wlUrl, URL url) {
```
Comment (hendrikebbers) [inline core/src/main/java/net/sourceforge/jnlp/util/whitelist/WhitelistEntry.java:66]: please define better names for the parameters

File: core/src/test/java/net/sourceforge/jnlp/util/whitelist/WhitelistEntryTest.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        assertMatch(validEntry, "http://test.com:1000");
+        assertMatch(validEntry, "http://test.com:10000");
+    }
+
+    @Test
+    public void wildcardHostPartShouldMatchAnyPrefix() throws Exception {
+        final WhitelistEntry validEntry = validateWhitelistUrl("http://*.test.com");
+
+        assertMatch(validEntry, "http://sub.test.com");
+        // FIXME: assertMatch(validEntry, "http://sub.sub.test.com");
```
Comment (sclassen) [inline core/src/test/java/net/sourceforge/jnlp/util/whitelist/WhitelistEntryTest.java:78]: @hendrikebbers 
Any comment on this FIXME?

Comment (hendrikebbers) [reply core/src/test/java/net/sourceforge/jnlp/util/whitelist/WhitelistEntryTest.java:78]: we just need to document that only 1 subdomain is supported for now. @janakmulani is this part of the doc?

File: core/src/main/java/net/adoptopenjdk/icedteaweb/client/controlpanel/panels/ServerWhitelistPanel.java [hunk 6]
```diff
@@ -46,9 +58,92 @@ public ServerWhitelistPanel(final DeploymentConfiguration config) {
...
+        table.setAutoResizeMode(JTable.AUTO_RESIZE_ALL_COLUMNS);
+        table.setAutoCreateRowSorter(true);
+
+        final TableColumnModel colModel = table.getColumnModel();
+        colModel.getColumn(0).setPreferredWidth(100);
+        colModel.getColumn(1).setPreferredWidth(250);
+
+        final DefaultTableCellRenderer cellRenderer = new EffectiveWhitelistCellRenderer();
+        table.getColumnModel().getColumn(1).setCellRenderer(cellRenderer);
+        cellRenderer.setIconTextGap(5);
```
Comment (hendrikebbers) [inline core/src/main/java/net/adoptopenjdk/icedteaweb/client/controlpanel/panels/ServerWhitelistPanel.java:75]: Let's add this to the EffectiveWhitelistCellRenderer constructor

File: core/src/main/java/net/adoptopenjdk/icedteaweb/client/controlpanel/panels/ServerWhitelistPanel.java [hunk 7]
```diff
@@ -46,9 +58,92 @@ public ServerWhitelistPanel(final DeploymentConfiguration config) {
...
+        cellRenderer.setVerticalTextPosition(SwingConstants.CENTER);
```
Comment (hendrikebbers) [inline core/src/main/java/net/adoptopenjdk/icedteaweb/client/controlpanel/panels/ServerWhitelistPanel.java:76]: Let's add this to the EffectiveWhitelistCellRenderer constructor

File: core/src/test/java/net/sourceforge/jnlp/util/whitelist/WhitelistEntryTest.java (hunk 5 above)
Comment (janakmulani) [reply core/src/test/java/net/sourceforge/jnlp/util/whitelist/WhitelistEntryTest.java:78]: yes

**Pull request merged**